### PR TITLE
Feature: Block pickaxe ability  on private island

### PR DESCRIPTION
**Pull request by Thunderblade73**
## Dependencies
 - #811 
# What
Blocks the mining ability on the private island. Main use is to prevent people to make cheese out of there island
## Discord: 
https://discord.com/channels/997079228510117908/1187705114874617947

### CODE REVIEW

File: .github/workflows/build.yml
Comment (lineargraph) [inline .github/workflows/build.yml:1]: Are these format changes relevant to this PR? (And with which reason are those format changes being applied?)
The stacktrace is also already available inside of the test report.

File: src/main/java/at/hannibal2/skyhanni/data/ItemClickData.kt [hunk 1]
```diff
@@ -23,18 +23,18 @@ class ItemClickData {
...
-                ItemClickEvent(InventoryUtils.getItemInHand(), ClickType.RIGHT_CLICK).postAndCatch()
+                event.isCanceled = ItemClickEvent(InventoryUtils.getItemInHand(), ClickType.RIGHT_CLICK).postAndCatch()
             }
         }
         if (packet is C07PacketPlayerDigging && packet.status == C07PacketPlayerDigging.Action.START_DESTROY_BLOCK) {
             val position = packet.position.toLorenzVec()
-            BlockClickEvent(ClickType.LEFT_CLICK, position, InventoryUtils.getItemInHand()).postAndCatch()
-            ItemClickEvent(InventoryUtils.getItemInHand(), ClickType.RIGHT_CLICK).postAndCatch()
+            event.isCanceled = BlockClickEvent(ClickType.LEFT_CLICK, position, InventoryUtils.getItemInHand()).postAndCatch() ||
+                ItemClickEvent(InventoryUtils.getItemInHand(), ClickType.RIGHT_CLICK).postAndCatch()
```
Comment (lineargraph) [inline src/main/java/at/hannibal2/skyhanni/data/ItemClickData.kt]: Consider cancelling the second event instead of using `||` which completely shortcircuits the second event:
```suggestion
            val blockClickCancelled = BlockClickEvent(ClickType.LEFT_CLICK, position, InventoryUtils.getItemInHand()).postAndCatch()
            event.isCanceled = ItemClickEvent(InventoryUtils.getItemInHand(), ClickType.RIGHT_CLICK).also { it.setCancelled(blockClickCancelled) }.postAndCatch()
```

This way `@SubscribeEvent(receiveCancelled = true)` will still get called.

File: src/main/java/at/hannibal2/skyhanni/data/RegexData.kt
Comment (lineargraph) [inline src/main/java/at/hannibal2/skyhanni/data/RegexData.kt:1]: I think regexes should be local to the module that uses them. Maybe create an API instead to cover these use cases for multiple files instead of directly exposing the regex.

File: src/main/java/at/hannibal2/skyhanni/utils/ItemUtils.kt [hunk 2]
```diff
@@ -205,29 +207,59 @@ object ItemUtils {
...
+                            "Could not read category for item $name",
+                            "Failed to read category from item rarity via item lore",
+                            "internal name" to getInternalName(),
+                            "item name" to name,
+                            "inventory name" to InventoryUtils.openInventoryName(),
+                            "pattern result" to itemCategory,
+                            "lore" to getLore(),
+                        )
+                    }
+                    return LorenzRarity.valueOf(rarity) to null
```
Comment (lineargraph) [inline src/main/java/at/hannibal2/skyhanni/utils/ItemUtils.kt]: This line will cause a crash if the regex changes later on to include more rarities. Either be wary of that by returning `null` for a rarity, or replace the `LorenzRarity` from an enum to a value class with an internal / private constructor.

File: .github/workflows/build.yml
Comment (Thunderblade73) [reply .github/workflows/build.yml:1]: The format change, idk. Intellji did them. And the stacktrace thing is from the depended PR from this comment: https://discord.com/channels/997079228510117908/1094190239532208228/1185953806555553882

File: src/main/java/at/hannibal2/skyhanni/data/RegexData.kt
Comment (Thunderblade73) [reply src/main/java/at/hannibal2/skyhanni/data/RegexData.kt:1]: Ehm, it was just a fix for the repoPattern and Test problem, where you couldn't (or didn't wanted) to give me an answer.

File: src/main/java/at/hannibal2/skyhanni/utils/ItemUtils.kt (hunk 2 above)
Comment (Thunderblade73) [reply src/main/java/at/hannibal2/skyhanni/utils/ItemUtils.kt]: thx, forgot it for the rarity

File: src/main/java/at/hannibal2/skyhanni/features/misc/PrivateIslandNoPickaxeAbility.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import at.hannibal2.skyhanni.utils.LorenzUtils.isInIsland
+import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
+
+class PrivateIslandNoPickaxeAbility {
+
+    val config get() = SkyHanniMod.feature.mining
+
+    @SubscribeEvent
+    fun onClick(event: WorldClickEvent) {
+        if (!LorenzUtils.inSkyBlock || !IslandType.PRIVATE_ISLAND.isInIsland()) return
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/misc/PrivateIslandNoPickaxeAbility.kt]: isInIsland already checks for inSkyBlock. the additional check can be removed

File: src/main/java/at/hannibal2/skyhanni/utils/LorenzUtils.kt [hunk 4]
```diff
@@ -141,7 +141,7 @@ object LorenzUtils {
 
     fun getSBMonthByName(month: String): Int {
         var monthNr = 0
-        for (i in 1..12) {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/utils/LorenzUtils.kt]: Kotlin does not want empty spaces before or after the .. operator. See the examples without space here: https://kotlinlang.org/docs/ranges.html

File: src/main/java/at/hannibal2/skyhanni/events/BlockClickEvent.kt [hunk 5]
```diff
@@ -4,7 +4,9 @@ import at.hannibal2.skyhanni.data.ClickType
 import at.hannibal2.skyhanni.utils.BlockUtils.getBlockStateAt
 import at.hannibal2.skyhanni.utils.LorenzVec
 import net.minecraft.item.ItemStack
+import net.minecraftforge.fml.common.eventhandler.Cancelable
 
+@Cancelable
 class BlockClickEvent(val clickType: ClickType, val position: LorenzVec, val itemInHand: ItemStack?) : LorenzEvent() {
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/events/BlockClickEvent.kt]: This does not extend world click event. is this intended?

**Review by hannibal002 — CHANGES_REQUESTED** [review]
right clicking on the island correctly blocks, but right clicking on blocks get sent to the server.

File: src/main/java/at/hannibal2/skyhanni/events/BlockClickEvent.kt (hunk 5 above)
Comment (Thunderblade73) [reply src/main/java/at/hannibal2/skyhanni/events/BlockClickEvent.kt]: that is not intended. Happend during merge conflict solving. (Also the bug with the block clicking)

**Review by hannibal002 — APPROVED** [review]
